src/Microsoft.AspNet.Mvc.Core/TempDataDictionary.cs

```diff
@@ -0,0 +1,272 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.AspNet.Hosting;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <inheritdoc />
+    public class TempDataDictionary : ITempDataDictionary
+    {
+        private Dictionary<string, object> _data;
+        private bool _loaded;
+        private ITempDataProvider _provider;
+        private IHttpContextAccessor _contextAccessor;
+        private HashSet<string> _initialKeys = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        private HashSet<string> _retainedKeys = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TempDataDictionary"/> class.
+        /// </summary>
+        /// <param name="context">The <see cref="IHttpContextAccessor"/> that provides the HttpContext.</param>
+        /// <param name="provider">The <see cref="ITempDataProvider"/> used to Load and Save data.</param>
+        public TempDataDictionary([NotNull] IHttpContextAccessor context, [NotNull] ITempDataProvider provider)
+        {
+            _provider = provider;
+            _loaded = false;
+            _contextAccessor = context;
+        }
+
+        public int Count
+        {
+            get
+            {
+                Load();
+                return _data.Count;
+            }
+        }
+
+        public ICollection<string> Keys
+        {
+            get
+            {
+                Load();
+                return _data.Keys;
+            }
+        }
+
+        public ICollection<object> Values
+        {
+            get
+            {
+                Load();
+                return _data.Values;
+            }
+        }
+
+        bool ICollection<KeyValuePair<string, object>>.IsReadOnly
+        {
+            get
+            {
+                Load();
+                return ((ICollection<KeyValuePair<string, object>>)_data).IsReadOnly;
+            }
+        }
+
+        public object this[string key]
+        {
+            get
+            {
+                Load();
+                object value;
+                if (TryGetValue(key, out value))
+                {
+                    _initialKeys.Remove(key);
+                    return value;
+                }
+                return null;
+            }
+            set
+            {
+                Load();
+                _data[key] = value;
+                _initialKeys.Add(key);
+            }
+        }
+
+        /// <inheritdoc />
+        public void Keep()
+        {
+            Load();
+            _retainedKeys.Clear();
+            _retainedKeys.UnionWith(_data.Keys);
```

Comment:
Oh wait duh I thought this was LINQ, didn't realize it's a HashSet-specific method. Ignore me.